src/lib/shim/src/tls.rs

```diff
@@ -0,0 +1,806 @@
+                0
+            );
+            // SAFETY: The TLS bytes for each thread are initialized to 0, and
+            // all-zeroes is a valid value of `ShimTlsVarStorage<T>`.
+            //
+            // We've ensure proper alignment when calculating the offset,
+            // and verified in the assertion just above.
+            let this_var: &ShimTlsVarStorage<T> = unsafe { &*this_var };
+            this_var
+        };
+
+        // SAFETY: This thread has exclusive access to the underlying storage,
+        // and we don't let references to `initd` escape this method,
+        // so we have exclusive access.
+        let initd: &mut i8 = unsafe { &mut *this_var.initd.get() };
+        if *initd == 0 {
+            // Initialize the value.
+
+            // SAFETY: This thread has exclusive access to the underlying storage.
+            // This is the only place we ever construct a mutable reference to this
+            // value, and we know we've never constructed a reference before, since
+            // the data isn't initialized.
+            let value: &mut MaybeUninit<T> = unsafe { &mut *this_var.value.get() };
+            value.write((self.f)());
+            *initd = 1;
+        }
+
+        // SAFETY: We've ensured this value is initialized, and we know there
+        // are no exclusive references outside of the one-time initialization
+        // block above.
+        let val: &T = unsafe { (*this_var.value.get()).assume_init_ref() };
+
+        TLSVarRef {
+            val,
+            _phantom: core::marker::PhantomData,
+        }
+    }
+}
+
+pub struct TLSVarRef<'a, T> {
+    val: &'a T,
+
+    // Force to be !Sync and !Send.
+    _phantom: core::marker::PhantomData<*mut T>,
+}
+// Double check `!Send` and `!Sync`.
+static_assertions::assert_not_impl_any!(TLSVarRef<'static, ()>: Send, Sync);
+
+impl<'a, T> Deref for TLSVarRef<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.val
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use core::cell::RefCell;
+    use core::sync::atomic::AtomicI16;
+
+    use vasi_sync::sync::AtomicI32;
+
+    use super::*;
+
+    fn test_each_mode(f: impl Fn()) {
+        // Run in each mode. We need to serialize here since the mode is global,
+        // and can't be safely changed while TLS from another mode is in use.
+        static MODE_MUTEX: std::sync::Mutex<()> = std::sync::Mutex::new(());
+        for mode in &[Mode::Native, Mode::NativeTlsId, Mode::Gettid] {
+            let _lock = MODE_MUTEX.lock().unwrap();
+            unsafe { global_preferred_mode::set(*mode) };
+            f();
+            unsafe { global_preferred_mode::reset_for_testing() };
+        }
+    }
+
+    #[test_log::test]
+    fn test_minimal() {
+        static MY_VAR: ShimTlsVar<u32> = ShimTlsVar::new(|| 0);
+        test_each_mode(|| {
+            assert_eq!(*MY_VAR.get(), 0);
+            unsafe { unregister_current_thread() };
+        });
+    }
```

Comment:
All of these tests will be running in parallel, so I don't see how the `test_each_mode` works properly.

Reply:
Oops I missed the mutex here somehow.